packages/apollo-engine-reporting/package.json

```diff
@@ -14,6 +14,7 @@
     "apollo-engine-reporting-protobuf": "file:../apollo-engine-reporting-protobuf",
     "apollo-graphql": "^0.3.3",
     "apollo-server-env": "file:../apollo-server-env",
+    "apollo-server-caching": "file:../apollo-server-caching",
```

Comment:
Since it's a type that's only used within a `private` property on the `export`ed `EngineReportingAgent` class, would this be better categorized within `devDependencies`?:

https://github.com/apollographql/apollo-server/blob/6f5769c441b4c87ea85fe285a78f5677a5310df6/packages/apollo-engine-reporting/src/agent.ts#L224-L225

(i.e. Could you verify to see if having it as a dev-dep resolves the concern you identified?)

Reply:
It does not, as we are only use dependencies when packaging for deploy; and using `devDependencies` would not be correct per npm recommendations:
> - **"dependencies"**: Packages required by your application in production.
> - **"devDependencies"**: Packages that are only needed for local development and testing.

~ See [Specifying dependencies and devDependencies in a package.json file](https://docs.npmjs.com/specifying-dependencies-and-devdependencies-in-a-package-json-file).

Reply:
To be clear, I'm not saying that this dependency doesn't need to be added, but the rational for deciding whether types go into `devDependencies` or `dependencies` is not as straightforward as those npm recommendations you've listed.

Concretely, if we went by the npm standard that you're stating here — which I'm familiar with — we could arguably say that no types are necessary since normal dependencies since TypeScript is compiled to JavaScript before it runs in production and the type definitions are only used during development.

You're probably aware that typically, `@types/*` packages are installed as `devDependencies`, but I'm curious, did you test my suggestion or are you just trying to be educational?  I'm not trying to be rude — and a reproduction of the problem or at least the error message would have made the solution more clear — but in the case of a library that provides typings (like `apollo-engine-reporting`, in this case!), it _is_ important to put types into `dependencies` if those types appear in the emitted definition files (i.e. `*.d.ts`) since consuming packages need to be able to access those definitions during consuming application's compilation. However, in the case of `InMemoryLRUCache` within `apollo-engine-reporting`, again, its only used in a `private` property on an exported `EngineReportingAgent` class declaration.

That means that `InMemoryLRUCache` _doesn't appear_ in [the emitted types within `agent.d.ts`](https://cdn.jsdelivr.net/npm/apollo-engine-reporting@1.4.2/dist/agent.d.ts), which merely notes it as a `private` property. 
 Since it doesn't appear there, the `InMemoryLRUCache` type shouldn't be necessary.

(Btw, if it were not `private`, you're absolutely correct that it should be in `dependencies`, as it would be present in the `agent.d.ts` file!

Reply:
Really, that's all to say, can you provide a reproduction or the error output? 😄

Reply:
Starting with the error:
```
Error: Cannot find module 'apollo-server-caching'
```
The stacktrace points to the transpiled version of: https://github.com/apollographql/apollo-server/blob/6f5769c441b4c87ea85fe285a78f5677a5310df6/packages/apollo-engine-reporting/src/agent.ts#L16 

##

The issue is with how npm handles version incompatibilities when building a package-lock.

We are in a situation where this dependency is not inlined in the `package-lock` but the other dependency we use that requires `apollo-server-caching` is inlined not allowing this package to access the inlined version of `apollo-server-caching`.

Since npm does not install `devDependencies` in this case, adding to `devDependencies` does not cause the package to get installed in the correct place.

I do not want to include my entire `package-lock.json` so I created a paired down example based on it, https://gist.github.com/squirly/1452f6f1dbec894da41293b0b3713e06. It is contrived but exposes the issue.

If you run `npm ci`; then try to require this module, `node -r apollo-server-caching`, you will get the above error.

This example also exposed that `graphql` should be a `dependency` or `peerDependency`.